server/test/auth/ApiAuthenticatorTest.java

```diff
@@ -0,0 +1,173 @@
+    // which makes it much less easy to inject things that are annotated with annotations
+    // that take arguments, such as @NamedCache("api-cache'). Guice makes that fairly
+    // though, hence this test class uses the guice injector directly.
+    GuiceApplicationBuilder builder =
+        new GuiceApplicationLoader().builder(new ApplicationLoader.Context(Environment.simple()));
+    Injector injector = Guice.createInjector(builder.applicationModule());
+    injector.injectMembers(this);
+
+    resourceCreator =
+        new ResourceCreator(
+            new play.inject.Injector() {
+              @Override
+              public <T> T instanceOf(Class<T> clazz) {
+                return injector.getInstance(clazz);
+              }
+
+              @Override
+              public <T> T instanceOf(BindingKey<T> key) {
+                return null;
+              }
+
+              @Override
+              public play.api.inject.Injector asScala() {
+                return null;
+              }
+            });
+
+    Helpers.start(application);
+
+    DB.getDefault().truncate("api_keys");
+    apiKey = resourceCreator.createActiveApiKey("test-key", keyId, secret);
+    cacheApi.remove(keyId);
+  }
+
+  @After
+  public void tearDown() {
+    Helpers.stop(application);
+  }
+
+  @Test
+  public void validate_success() {
+    apiAuthenticator.validate(
+        new UsernamePasswordCredentials(keyId, secret),
+        new PlayWebContext(buildFakeRequest(validRawCredentials)),
+        MOCK_SESSION_STORE);
+
+    Optional<Optional<ApiKey>> cacheEntry = cacheApi.get(keyId);
+    Optional<ApiKey> cachedMaybeKey = cacheEntry.get();
+    assertThat(cachedMaybeKey.get().id).isEqualTo(apiKey.id);
+  }
+
+  @Test
+  public void validate_invalidKeyId() {
+    String rawCredentials = "wrong" + ":" + secret;
+
+    assertBadCredentialsException(
+        buildFakeRequest(rawCredentials),
+        new UsernamePasswordCredentials("wrong", secret),
+        "API key does not exist");
+  }
+
+  @Test
+  public void validate_retiredKey() {
+    apiKey.retire("test-user");
+    apiKey.save();
+
+    assertBadCredentialsException(buildFakeRequest(validRawCredentials), "API key is retired");
+  }
+
+  @Test
+  public void validate_isExpired() {
+    apiKey.setExpiration(Instant.now().minusSeconds(100));
+    apiKey.save();
+
+    assertBadCredentialsException(buildFakeRequest(validRawCredentials), "API key is expired");
+  }
+
+  @Test
+  public void validate_ipNotInSubnet() {
+    apiKey.setSubnet("2.2.2.2/30");
+    apiKey.save();
+
+    assertBadCredentialsException(buildFakeRequest(validRawCredentials), "IP not in allowed range");
+  }
+
+  @Test
+  public void validate_invalidSecret() {
+    var rawCredentials = keyId + ":" + "notthesecret";
+
+    assertBadCredentialsException(
+        buildFakeRequest(rawCredentials),
+        new UsernamePasswordCredentials(keyId, "notthesecret"),
+        "Invalid secret");
```

Comment:
I would think that we should just return an invalid key response to this case (and maybe some of the others) - no need to give extra information to someone who may be trying to break into the system

Reply:
The message in these exceptions isn't included in the response, just the server logs:

```
*   Trying 127.0.0.1:9000...
* Connected to localhost (127.0.0.1) port 9000 (#0)
> GET /api/v1/admin/programs/mock-program/applications HTTP/1.1
> Host: localhost:9000
> User-Agent: curl/7.81.0
> Accept: */*
> Authorization: Basic TTVBK25FVlUyMlZzbjdPSFF2QWZIQT09OlVWSVAzd0ExZFBaMDhSNFZSTjY5djFOTExKUVFlTnBvQXQ5N0xjU3A4ZUE9
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< Date: Tue, 17 May 2022 23:32:48 GMT
< Content-Length: 0
< 
* Connection #0 to host localhost left intact
```

Reply:
👍

Reply:
Perhaps we can add that to the validate function documentation